### PR TITLE
New versioning system [preview nightly]

### PR DESCRIPTION
**Pull request by bendk**
**Merge note:** Most of these changes are approved, but we can't merge it because of blockers from other repos.  In fact, almost all of these PRs need to be approved and merged at the same time.  We don't want to update the firefox-android/firefox-ios code to expect the new versioning system without also merging this PR.  Here's a summary of the work:

- [x] Add taskcluster task to build the swift code (this is the one remaining code task for this repo)
  - [x] blocked by **relops 1821390**  [Add darwin worker for application-services](https://bugzilla.mozilla.org/show_bug.cgi?id=1821390)
- [ ] **firefox-android 1823462** [New application-services release system](https://github.com/mozilla-mobile/firefox-android/pull/1325)
- [x] **relbot#105** [Work with the new application-services release process](https://github.com/mozilla-mobile/relbot/pull/105)
- [x] **rust-components-swift#98** [New app services release process](https://github.com/mozilla/rust-components-swift/pull/98)
- [x] **firefox-ios #13806:** [Refactor app-services update code](https://github.com/mozilla-mobile/firefox-ios/pull/13806)
---

- Update the version number to `112.0.a1`.  It now matches the firefox version as discussed in ADR-0006.
- Update the nightly build task to replace the last component of that version number with a timestamp.
- Place nightly builds in a different directory in the Maven repository, like how geckoview uses a different directory for the nightly channel.

I think this is how we're going to want to handle our nightlies going forward, but I would love feedback from releng/relman.

We're hoping to move to using ship-it for cutting releases at the end of the nightly cycle and uplifts.  Should we be using ship-it for nightlies as well, or can we just continue to use the taskcluster cron task?  Is there anything that should change to be more compatible with ship-it?

Is there any way to test running the nightly taskcluster task from this branch?
 
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[ff-android: firefox-android-branch-name]` and/or `[fenix: fenix-branch-name]` to the PR title.

### CODE REVIEW

File: taskcluster/app_services_taskgraph/transforms/__init__.py
Comment (bendk) [inline taskcluster/app_services_taskgraph/transforms/__init__.py:36]: Is this the way we should handle nightlies?  I used `geckoview` as the guide here.  It has a [different directory for each channel in maven](https://maven.mozilla.org/?prefix=maven2/org/mozilla/geckoview/).

I think other components use a different maven repo and/or package name.  Should we be doing that?

Should normal releases go into a separate directory as well?

**Comment by rvandermeulen** [comment]
You'll want a version number of `112.0a1` rather than `112.0.0a1` IMO. And I think you should stick to using a cron for Nightly builds rather than doing so in Ship-It. Seems needlessly complicated otherwise.

For the rest, I'll defer to ahal & jlorenzo 😄

**Comment by bendk** [comment]
That makes sense, I updated the version number to match.  I also realized that the task filtering code wouldn't work at all so I updated it.  I would really love a way to test it for real though.

File: taskcluster/app_services_taskgraph/__init__.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/__init__.py:45]: Noting that technically this is a `kind_config` not a `graph_config`. But the misnomer is pre-existing so not something you need to fix here.

File: taskcluster/app_services_taskgraph/build_config.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/build_config.py:34]: Looks like only `config.params` is ever used, so might be better to make this function just accept `params`?

File: taskcluster/app_services_taskgraph/build_config.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/build_config.py:45]: nitpick: doesn't feel worth a separate function to me, but up to you :)

**Review by ahal — APPROVED** [review]
Looks good to me, just a few minor nits.

I don't know that I'm the best reviewer here, so I'd feel more comfortable if you also blocked on Johan's review.

File: taskcluster/app_services_taskgraph/target_tasks.py
Comment (JohanLorenzo) [inline taskcluster/app_services_taskgraph/target_tasks.py:13]: I've never tried setting this decorator to the same function. I wonder if it works as intended 🤔

File: taskcluster/app_services_taskgraph/transforms/__init__.py
Comment (JohanLorenzo) [reply taskcluster/app_services_taskgraph/transforms/__init__.py:36]: Good question! 

I know Android-Components are shipped on [nightly.maven.mozilla.org](https://nightly.maven.mozilla.org/?prefix=maven2/org/mozilla/components/). We've never moved geckoview there likely because nightlies.m.o arrived after we first shipped geckoview nightly.

File: taskcluster/app_services_taskgraph/__init__.py
Comment (JohanLorenzo) [inline taskcluster/app_services_taskgraph/__init__.py:64]: Nit: I wouldn't force this value here. `cron` decision task already set `target_tasks_method`.

**Review by JohanLorenzo — APPROVED** [review]
Looks good to me modulo a few minor things!

File: taskcluster/app_services_taskgraph/__init__.py
Comment (bendk) [reply taskcluster/app_services_taskgraph/__init__.py:64]: I'll remove the line.  Can you point to where that happens so I can understand better?

**Comment by bendk** [comment]
Thanks for all the help with this.  I just pushed 2 changes: one to address the review and one that:

- Allows nightlies to be triggered based on the PR title.  I'm going to test that with PR, let's hope I don't break anything :crossed_fingers: 
- Extends the concept of "nightly" to "preview build".  Preview builds can either be nightlies or a build from a PR.  The way I'm picturing this working is:
  - I have an app-services PR with breaking changes, so I make a preview build
  - I then open PRs on `firefox-android` and `firefox-ios` which use the app-services packages built by the preview build.  This allows reviews to see if the CI is green or not
  - If everything looks good, and the reviewers approve, I merge the app-services PR into main.
  - I don't merge the consumer app PRs directly, instead I wait for the next app-services nightly, then merge them into the PR branch for that nightly.
- Upload nightly/preview builds to the staging maven bucket.

File: taskcluster/app_services_taskgraph/__init__.py
Comment (JohanLorenzo) [reply taskcluster/app_services_taskgraph/__init__.py]: Sure! I have some good news, the location is already in your repo, it's here: 

https://github.com/mozilla/application-services/blob/3c4c626295d098d67fe2fa34f51a5552b6d441cc/.cron.yml#L11

**Review by jhugman — APPROVED** [review]
This looks great, Ben. I've tested the `nimbus-fml` and this seems to work just great.

Excellent!

File: taskcluster/ci/nimbus-fml/kind.yml
Comment (bendk) [inline taskcluster/ci/nimbus-fml/kind.yml:55]: This commit moves some of our artifact storage from github releases to taskcluster.  How long are indexed task artifacts stored?
  - I'm hoping that the nightly artifacts would go back at least 6 months
  - I'm hoping release artifacts would never expire.  Is that possible with taskcluster?  If not, I'm thinking we could do a github release for those.  Are there any other options?

@ahal @JohanLorenzo

Comment (bendk) [reply taskcluster/ci/nimbus-fml/kind.yml:55]: I think I've figured this out:
  - [All indexed entities must have an expiration](https://docs.taskcluster.net/docs/reference/core/index#entry-expiration).  I think this means we should do a github release for our non-nightly releases.
  - Our nightly releases are currently set to expire after 28 days.  I'll bump this to 180.

Comment (ahal) [reply taskcluster/ci/nimbus-fml/kind.yml:55]: Could you provide a link to the task please?

Artifact expiry and task expiry are distinct from one another, so one possibility is that you are looking at the artifact expiry locally. Another possibility is that the expiry value is being overwritten by some later phase of the generation (e.g a morph?), though I'm not aware of anything that should be doing this off the top of my head.

Comment (ahal) [reply taskcluster/ci/nimbus-fml/kind.yml:55]: This took me way too long to track down, but these are the Taskgraph defaults for try (or pull requests) vs pushes respectively:
https://github.com/taskcluster/taskgraph/blob/main/src/taskgraph/transforms/task.py#L1027

So I think everything is working as expected? You can add an `expires-after` key to the task definition if you'd like to have your own custom expiry / behaviour.

Comment (ahal) [reply taskcluster/ci/nimbus-fml/kind.yml:55]: Specifically the parameter file you used locally has `tasks_for` set to `cron`, and that `is_try()` function only returns True if it's set to `github-pull-request`:
https://github.com/taskcluster/taskgraph/blob/main/src/taskgraph/parameters.py#L226

**Pull request merged**
